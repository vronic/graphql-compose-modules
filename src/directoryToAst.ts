import fs from 'fs';
import { join, resolve, dirname, basename } from 'path';

export interface DirectoryToAstOptions {
  relativePath?: string;
  extensions?: string[];
  include?: RegExp | ((path: string, kind: 'dir' | 'file', filename: string) => boolean);
  exclude?: RegExp | ((path: string, kind: 'dir' | 'file', filename: string) => boolean);
}

export type AstNodeKinds = 'rootType' | 'dir' | 'file' | 'root';

export interface AstBaseNode {
  kind: AstNodeKinds;
  name: string;
  absPath: string;
}

export interface AstRootTypeNode extends AstBaseNode {
  kind: 'rootType';
  children: AstDirChildren;
  namespaceConfig?: AstFileNode;
}

export type AstDirChildren = {
  [key: string]: AstDirNode | AstFileNode;
};

export interface AstDirNode extends AstBaseNode {
  kind: 'dir';
  children: AstDirChildren;
  namespaceConfig?: AstFileNode;
}

export interface AstFileNode extends AstBaseNode {
  kind: 'file';
  code: {
    default?: any;
  };
}

export type RootTypeNames = 'query' | 'mutation' | 'subscription';

export interface AstRootNode extends AstBaseNode {
  kind: 'root';
  children: {
    [T in RootTypeNames]?: AstRootTypeNode;
  };
}

export const defaultOptions: DirectoryToAstOptions = {
  extensions: ['js', 'ts'],
};

export function directoryToAst(
  m: NodeModule,
  options: DirectoryToAstOptions = defaultOptions
): AstRootNode {
  // if no path was passed in, assume the equivelant of __dirname from caller
  // otherwise, resolve path relative to the equivalent of __dirname
  const schemaPath = options?.relativePath
    ? resolve(dirname(m.filename), options.relativePath)
    : dirname(m.filename);

  // setup default options
  Object.keys(defaultOptions).forEach((prop) => {
    if (typeof (options as any)[prop] === 'undefined') {
      (options as any)[prop] = (defaultOptions as any)[prop];
    }
  });

  const result = {
    kind: 'root',
    name: basename(schemaPath),
    absPath: schemaPath,
    children: {},
  } as AstRootNode;

  fs.readdirSync(schemaPath).forEach((filename) => {
    const absPath = join(schemaPath, filename);

    if (fs.statSync(absPath).isDirectory()) {
      const dirName = filename;
      const re = /^(query|mutation|subscription)(\.(.*))?$/i;
      const found = dirName.match(re);
      if (found) {
        const opType = found[1].toLowerCase() as keyof AstRootNode['children'];
        let rootTypeAst = result.children[opType];
        if (!rootTypeAst)
          rootTypeAst = {
            kind: 'rootType',
            name: opType,
            absPath,
            children: {},
          } as AstRootTypeNode;

        const astDir = getAstForDir(m, absPath, options);
        if (astDir) {
          const subField = found[3]; // any part after dot (eg for `query.me` will be `me`)
          if (subField) {
            rootTypeAst.children[subField] = {
              ...astDir,
              name: subField,
              absPath,
            };
          } else {
            rootTypeAst.children = astDir.children;
            if (astDir.namespaceConfig) {
              rootTypeAst.namespaceConfig = astDir.namespaceConfig;
            }
          }
          result.children[opType] = rootTypeAst;
        }
      }
    }
  });

  return result;
}

export function getAstForDir(
  m: NodeModule,
  absPath: string,
  options: DirectoryToAstOptions = defaultOptions
): AstDirNode | void {
  const name = basename(absPath);

  if (!checkInclusion(absPath, 'dir', name, options)) return;

  const result: AstDirNode = {
    kind: 'dir',
    absPath,
    name,
    children: {},
  };

  // get the path of each file in specified directory, append to current tree node, recurse
  fs.readdirSync(absPath).forEach((filename) => {
    const absFilePath = join(absPath, filename);

    const stat = fs.statSync(absFilePath);
    if (stat.isDirectory()) {
      // this node is a directory; recurse
      if (result.children[filename]) {
        throw new Error(
          `You have a folder and file with same name "${filename}" by the following path ${absPath}. Please remove one of them.`
        );
      }
      const astDir = getAstForDir(m, absFilePath, options);
      if (astDir) {
        result.children[filename] = astDir;
      }
    } else if (stat.isFile()) {
      // this node is a file
      const fileAst = getAstForFile(m, absFilePath, options);
      if (fileAst) {
        if (fileAst.name === 'index') {
          result.namespaceConfig = fileAst;
        } else if (result.children[fileAst.name]) {
          throw new Error(
            `You have a folder and file with same name "${fileAst.name}" by the following path ${absPath}. Please remove one of them.`
          );
        } else {
          result.children[fileAst.name] = fileAst;
        }
      }
    }
  });

  return result;
}

export function getAstForFile(
  m: NodeModule,
  absPath: string,
  options: DirectoryToAstOptions = defaultOptions
): AstFileNode | void {
  const filename = basename(absPath);
  if (absPath !== m.filename && checkInclusion(absPath, 'file', filename, options)) {
    // module name shouldn't include file extension
    const moduleName = filename.substring(0, filename.lastIndexOf('.'));
    return {
      kind: 'file',
      name: moduleName,
      absPath,
      code: m.require(absPath),
    };
  }
}

function checkInclusion(
  absPath: string,
  kind: 'dir' | 'file',
  filename: string,
  options: DirectoryToAstOptions
): boolean {
  // Skip dir/files started from double underscore
  if (/^__.*/i.test(filename)) {
    return false;
  }

  // Skip dir/files started from dot
  if (/^\..*/i.test(filename)) {
    return false;
  }

  if (kind === 'file') {
    if (
      // Verify file has valid extension
      !new RegExp('\\.(' + (options?.extensions || ['js', 'ts']).join('|') + ')$', 'i').test(
        filename
      ) ||
      // Hardcoded skip file extensions
      // typescript definition files
      new RegExp('(\\.d\\.ts)$', 'i').test(filename) ||
      // test files
      new RegExp('(\\.spec\\.(ts|js))$', 'i').test(filename)
    ) {
      return false;
    }
  }

  if (options.include) {
    if (options.include instanceof RegExp) {
      // if options.include is a RegExp, evaluate it and make sure the path passes
      if (!options.include.test(absPath)) return false;
    } else if (typeof options.include === 'function') {
      // if options.include is a function, evaluate it and make sure the path passes
      if (!options.include(absPath, kind, filename)) return false;
    }
  }

  if (options.exclude) {
    if (options.exclude instanceof RegExp) {
      // if options.exclude is a RegExp, evaluate it and make sure the path doesn't pass
      if (options.exclude.test(absPath)) return false;
    } else if (typeof options.exclude === 'function') {
      // if options.exclude is a function, evaluate it and make sure the path doesn't pass
      if (options.exclude(absPath, kind, filename)) return false;
    }
  }

  return true;
}
