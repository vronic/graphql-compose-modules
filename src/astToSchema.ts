import {
  SchemaComposer,
  ObjectTypeComposer,
  upperFirst,
  ObjectTypeComposerFieldConfig,
  isOutputTypeDefinitionString,
  isWrappedTypeNameString,
  isComposeOutputType,
  isSomeOutputTypeDefinitionString,
  inspect,
} from 'graphql-compose';
import { AstResult, AstRootTypeNode, AstDirNode, AstFileNode } from './directoryToAst';
import dedent from 'dedent';
import { GraphQLObjectType } from 'graphql';

export interface AstToSchemaOptions {
  schemaComposer?: SchemaComposer<any>;
  prefix?: string;
  suffix?: string;
}

export function astToSchema<TContext = any>(
  ast: AstResult,
  opts: AstToSchemaOptions = {}
): SchemaComposer<TContext> {
  let sc: SchemaComposer<any>;

  if (opts?.schemaComposer) {
    if (!opts.schemaComposer) {
      throw new Error(dedent`
        Provided option 'schemaComposer' should be an instance of SchemaComposer class from 'graphql-compose' package. 
        Recieved:
          ${inspect(opts.schemaComposer)}
      `);
    }
    sc = opts.schemaComposer;
  } else {
    sc = new SchemaComposer();
  }

  if (ast.query) populateRoot(sc, 'Query', ast.query, opts);
  if (ast.mutation) populateRoot(sc, 'Mutation', ast.mutation, opts);
  if (ast.subscription) populateRoot(sc, 'Subscription', ast.subscription, opts);

  return sc;
}

function populateRoot(
  sc: SchemaComposer<any>,
  rootName: 'Query' | 'Mutation' | 'Subscription',
  astRootNode: AstRootTypeNode,
  opts?: AstToSchemaOptions
) {
  const tc = sc[rootName];
  Object.keys(astRootNode.children).forEach((key) => {
    createFields(sc, astRootNode.children[key], tc, rootName, opts || {});
  });
}

export function createFields(
  sc: SchemaComposer<any>,
  ast: AstDirNode | AstFileNode | void,
  parent: ObjectTypeComposer,
  pathPrefix: string,
  opts: AstToSchemaOptions = {}
): void {
  if (!ast) return;

  const name = ast.name;
  if (!/^[._a-zA-Z0-9]+$/.test(name)) {
    throw new Error(
      `You provide incorrect ${
        ast.kind === 'file' ? 'file' : 'directory'
      } name '${name}', it should meet RegExp(/^[._a-zA-Z0-9]+$/) for '${ast.absPath}'`
    );
  }

  if (ast.kind === 'file') {
    parent.addNestedFields({
      [name]: prepareFieldConfig(sc, ast),
    });
    return;
  }

  if (ast.kind === 'dir') {
    const typename = getTypename(ast, pathPrefix, opts);
    let fc: ObjectTypeComposerFieldConfig<any, any>;
    if (ast.namespaceConfig) {
      fc = prepareNamespaceFieldConfig(sc, ast.namespaceConfig, typename);
    } else {
      fc = { type: sc.createObjectTC(typename) };
    }

    parent.addNestedFields({
      [name]: {
        resolve: () => ({}),
        ...fc,
      },
    });

    const pathPrefixForChild = getTypename(ast, pathPrefix, {});
    Object.keys(ast.children).forEach((key) => {
      createFields(sc, ast.children[key], fc.type as any, pathPrefixForChild, opts);
    });
  }
}

function getTypename(
  ast: AstDirNode | AstFileNode,
  pathPrefix: string,
  opts: AstToSchemaOptions
): string {
  const name = ast.name;

  let typename = pathPrefix;
  if (name.indexOf('.') !== -1) {
    const namesArray = name.split('.');

    if (namesArray.some((n) => !n)) {
      throw new Error(
        `Field name '${ast.name}' contains dots in the wrong place for '${ast.absPath}'!`
      );
    }

    typename += namesArray.reduce((prev, current) => {
      return prev + upperFirst(current);
    }, '');
  } else {
    typename += upperFirst(name);
  }

  if (opts.prefix) typename = `${opts.prefix}${typename}`;
  if (opts.suffix) typename += opts.suffix;
  return typename;
}

function prepareNamespaceFieldConfig(
  sc: SchemaComposer<any>,
  ast: AstFileNode,
  typename: string
): ObjectTypeComposerFieldConfig<any, any> {
  if (!ast.code.default) {
    throw new Error(dedent`
      NamespaceModule MUST return FieldConfig as default export in '${ast.absPath}'. 
      Eg:
        export default {
          type: 'SomeObjectTypeName',
          resolve: () => Date.now(),
        };
    `);
  }

  const fc: any = ast.code.default;

  if (!fc.type) {
    fc.type = sc.createObjectTC(typename);
  } else {
    if (typeof fc.type === 'string') {
      if (!isOutputTypeDefinitionString(fc.type) && !isWrappedTypeNameString(fc.type)) {
        throw new Error(dedent`
          You provide incorrect output type definition:
            ${fc.type}
          It must be valid TypeName or output type SDL definition:
          
          Eg.
            type Payload { me: String }
          OR
            Payload
        `);
      }
    } else if (
      !(fc.type instanceof ObjectTypeComposer) &&
      !(fc.type instanceof GraphQLObjectType)
    ) {
      throw new Error(dedent`
        You provide some strange value as 'type':
          ${inspect(fc.type)}
      `);
    }
    fc.type = sc.createObjectTC(fc.type);
  }

  if (!fc.resolve) {
    fc.resolve = () => ({});
  }

  return fc;
}

function prepareFieldConfig(
  sc: SchemaComposer<any>,
  ast: AstFileNode
): ObjectTypeComposerFieldConfig<any, any> {
  const fc = ast.code.default as any;

  if (!fc) {
    throw new Error(dedent`
      Module MUST return FieldConfig as default export in '${ast.absPath}'. 
      Eg:
        export default {
          type: 'String',
          resolve: () => Date.now(),
        };
    `);
  }

  if (!fc.type || !isSomeOutputTypeDefinition(fc.type)) {
    throw new Error(dedent`
      Module MUST return FieldConfig with correct 'type: xxx' property in '${ast.absPath}'. 
      Eg:
        export default {
          type: 'String'
        };
    `);
  }

  return fc;
}

function isSomeOutputTypeDefinition(type: any): boolean {
  if (typeof type === 'string') {
    // type: 'String'
    return isSomeOutputTypeDefinitionString(type) || isWrappedTypeNameString(type);
  } else if (Array.isArray(type)) {
    // type: ['String']
    return isSomeOutputTypeDefinition(type[0]);
  } else {
    // type: 'type User { name: String }'
    return isComposeOutputType(type);
  }
}
