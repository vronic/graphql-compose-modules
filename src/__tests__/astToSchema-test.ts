import { directoryToAst, getAstForDir, getAstForFile } from '../directoryToAst';
import { astToSchema, createFields } from '../astToSchema';
import { SchemaComposer } from 'graphql-compose';
import { printSchema } from 'graphql/utilities';
import path from 'path';
import dedent from 'dedent';

describe('astToSchema()', () => {
  describe('Schema ./__testSchema__', () => {
    const ast = directoryToAst(module, { relativePath: './__testSchema__' });
    const sc = astToSchema(ast);

    it('should return schema composer', () => {
      expect(sc).toBeInstanceOf(SchemaComposer);
    });

    it('schema', async () => {
      const printedSchema = await printSchema(sc.buildSchema(), { commentDescriptions: true });
      expect(printedSchema).toMatchSnapshot();
    });
  });

  describe('createFields', () => {
    let sc: SchemaComposer<any>;
    beforeEach(() => {
      sc = new SchemaComposer();
    });

    it('file: query/field.ts', () => {
      createFields(
        sc,
        getAstForFile(module, path.resolve(__dirname, './__testSchema__/query/field.ts')),
        sc.Query,
        'Query'
      );
      expect(sc.Query.hasField('field')).toBeTruthy();
      expect(sc.Query.getFieldTypeName('field')).toBe('String');
    });

    it('file: query/some.nested.ts', () => {
      createFields(
        sc,
        getAstForFile(module, path.resolve(__dirname, './__testSchema__/query/some.nested.ts')),
        sc.Query,
        'Query'
      );
      expect(sc.Query.hasField('some')).toBeTruthy();
      expect(sc.Query.getFieldTypeName('some')).toBe('QuerySome');
      expect(sc.Query.getFieldOTC('some').hasField('nested')).toBeTruthy();
      expect(sc.Query.getFieldOTC('some').getFieldTypeName('nested')).toBe('Int');
    });

    it('file: query/some.index.ts should be as regular field', () => {
      createFields(
        sc,
        getAstForFile(module, path.resolve(__dirname, './__testSchema__/query/some.index.ts')),
        sc.Query,
        'Query'
      );
      expect(sc.Query.hasField('some')).toBeTruthy();
      expect(sc.Query.getFieldTypeName('some')).toBe('QuerySome');
      expect(sc.Query.getFieldOTC('some').hasField('index')).toBeTruthy();
      expect(sc.Query.getFieldOTC('some').getFieldTypeName('index')).toBe('SomeIndexFileType');
      expect((sc.Query.getFieldOTC('some').getFieldConfig('index') as any).resolve()).toEqual({
        awesomeValue: 'awesomeValue',
      });
    });

    it('dir: query/me/', () => {
      createFields(
        sc,
        getAstForDir(module, path.resolve(__dirname, './__testSchema__/query/me')),
        sc.Query,
        'Query'
      );
      expect(sc.Query.hasField('me')).toBeTruthy();
      expect(sc.Query.getFieldTypeName('me')).toBe('QueryMe');

      // check query/me/index.ts
      // should provide args for `me` field
      expect(sc.Query.getFieldArgTypeName('me', 'arg')).toBe('Int');
      expect(() => (sc.Query.getFieldConfig('me') as any).resolve()).toThrow(
        'You should be the ADMIN'
      );
      // check that fields from sibling files was added
      expect(sc.Query.getFieldOTC('me').getFieldTypeName('name')).toBe('String');
      expect((sc.Query.getFieldOTC('me').getFieldConfig('name') as any).resolve()).toBe('nodkz');
      expect(
        sc.Query.getFieldOTC('me')
          .getFieldOTC('address')
          .getTypeName()
      ).toBe('QueryMeAddress');
      expect(
        sc.Query.getFieldOTC('me')
          .getFieldOTC('address')
          .getFieldNames()
          .sort()
      ).toEqual(['city', 'street']);
    });
  });

  it('should properly set name for nested fields with dot notation', () => {
    const ast = directoryToAst(module, {
      relativePath: './__testSchema__',
      include: /query\.auth$|query\.auth\/nested/,
    });
    const sc = astToSchema(ast);
    expect(sc.Query.getFieldOTC('auth').getFieldTypeName('nested')).toBe('QueryAuthNested');
    expect(sc.Query.toSDL({ deep: true })).toBe(dedent`
      type Query {
        auth: QueryAuth
      }

      type QueryAuth {
        nested: QueryAuthNested
      }
      
      type QueryAuthNested {
        method: Boolean
      }
      
      """The \`Boolean\` scalar type represents \`true\` or \`false\`."""
      scalar Boolean
    `);
  });

  describe('astToSchema()', () => {
    it('should properly add prefix for created TypeNames', () => {
      const ast = directoryToAst(module, {
        relativePath: './__testSchema__',
        include: /query\.auth$|query\.auth\/nested/,
      });
      const sc = astToSchema(ast, { prefix: 'Corp', suffix: 'Old' });
      expect(sc.Query.toSDL({ deep: true })).toBe(dedent`
        type Query {
          auth: CorpQueryAuthOld
        }

        type CorpQueryAuthOld {
          nested: CorpQueryAuthNestedOld
        }
        
        type CorpQueryAuthNestedOld {
          method: Boolean
        }

        """The \`Boolean\` scalar type represents \`true\` or \`false\`."""
        scalar Boolean
      `);
    });
  });
});
