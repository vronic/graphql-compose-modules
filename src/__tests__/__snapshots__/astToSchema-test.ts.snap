// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`astToSchema() Schema ./__testSchema__ schema 1`] = `
"# Provides default value for input field.
directive @default(value: JSON!) on INPUT_FIELD_DEFINITION

# The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
scalar JSON

type Mutation {
  logs: MutationLogs
  auth: MutationAuth
  user: MutationUser
}

type MutationAuth {
  login: Boolean
  logout: Boolean
  nested: MutationAuthNested
}

type MutationAuthNested {
  method: Boolean
}

type MutationLogs {
  nested: MutationLogsNested
}

type MutationLogsNested {
  list: Boolean
}

type MutationUser {
  create: Boolean
  update: Boolean
}

type Query {
  field: String
  me(arg: Int): QueryMe
  some: QuerySome
  user: UserAwesomeType
  auth: QueryAuth
}

type QueryAuth {
  isLoggedIn: Boolean
  nested: QueryAuthNested
}

type QueryAuthNested {
  method: Boolean
}

type QueryMe {
  address: QueryMeAddress
  name: String
}

type QueryMeAddress {
  city: Boolean
  street: Boolean
}

type QuerySome {
  index: SomeIndexFileType
  nested: Int
}

type SomeIndexFileType {
  awesomeValue: String
}

type UserAwesomeType {
  firstName: String
  lastName: String
  extendedData: UserExtendedData
  roles: [String]
}

type UserExtendedData {
  starsCount: Int
}
"
`;
